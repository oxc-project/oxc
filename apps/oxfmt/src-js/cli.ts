#!/usr/bin/env node

import { format } from './bindings.js';
import { formatEmbeddedCode } from './embedded.js';

const args = process.argv.slice(2);

// Call the Rust formatter with our JS callback
const success = await format(args, formatEmbeddedCode);

process.exit(success ? 0 : 1);
