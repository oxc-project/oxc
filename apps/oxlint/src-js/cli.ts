import { createRequire } from 'node:module';
import { lint } from './bindings.js';

// Lazy-load `loadPlugin` and `lintFile` functions, on first call to `loadPlugin`.
// This avoids loading this code if user doesn't utilize JS plugins.
let loadPlugin: typeof loadPluginWrapper | null = null;
let lintFile: typeof lintFileWrapper | null = null;

function loadPluginWrapper(path: string): Promise<string> {
  if (loadPlugin === null) {
    const require = createRequire(import.meta.url);
    ({ loadPlugin, lintFile } = require('./plugins/index.js'));
  }
  return loadPlugin(path);
}

function lintFileWrapper(filePath: string, bufferId: number, buffer: Uint8Array | null, ruleIds: number[]): string {
  // `lintFile` is never called without `loadPlugin` being called first, so `lintFile` must be defined here
  return lintFile(filePath, bufferId, buffer, ruleIds);
}

// Call Rust, passing `loadPlugin` and `lintFile` as callbacks
const success = await lint(loadPluginWrapper, lintFileWrapper);

// Note: It's recommended to set `process.exitCode` instead of calling `process.exit()`.
// `process.exit()` kills the process immediately and `stdout` may not be flushed before process dies.
// https://nodejs.org/api/process.html#processexitcode
if (!success) process.exitCode = 1;
