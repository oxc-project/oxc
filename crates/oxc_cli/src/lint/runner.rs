use std::{
    fs,
    io::{BufWriter, Write},
    path::{Path, PathBuf},
    rc::Rc,
    sync::{
        atomic::{AtomicUsize, Ordering},
        mpsc, Arc,
    },
};

use miette::NamedSource;
use oxc_allocator::Allocator;
use oxc_ast::SourceType;
use oxc_diagnostics::{Error, GraphicalReportHandler, MinifiedFileError, Severity};
use oxc_linter::{Fixer, Linter, RuleCategory, RuleEnum, RULES};
use oxc_parser::Parser;
use oxc_semantic::SemanticBuilder;
use rustc_hash::FxHashSet;

use super::{AllowWarnDeny, LintOptions};
use crate::{CliRunResult, Walk};

pub struct LintRunner {
    options: LintOptions,

    linter: Arc<Linter>,
}

impl LintRunner {
    #[must_use]
    pub fn new(options: LintOptions) -> Self {
        let linter = Linter::from_rules(Self::derive_rules(&options)).with_fix(options.fix);
        Self { options, linter: Arc::new(linter) }
    }

    fn derive_rules(options: &LintOptions) -> Vec<RuleEnum> {
        let mut rules: FxHashSet<RuleEnum> = FxHashSet::default();

        for (allow_warn_deny, name_or_category) in &options.rules {
            let maybe_category = RuleCategory::from(name_or_category.as_str());
            match allow_warn_deny {
                AllowWarnDeny::Deny => {
                    match maybe_category {
                        Some(category) => rules.extend(
                            RULES.iter().filter(|rule| rule.category() == category).cloned(),
                        ),
                        None => {
                            if name_or_category == "all" {
                                rules.extend(RULES.iter().cloned());
                            } else {
                                rules.extend(
                                    RULES
                                        .iter()
                                        .filter(|rule| rule.name() == name_or_category)
                                        .cloned(),
                                );
                            }
                        }
                    };
                }
                AllowWarnDeny::Allow => {
                    match maybe_category {
                        Some(category) => rules.retain(|rule| rule.category() != category),
                        None => {
                            if name_or_category == "all" {
                                rules.drain();
                            } else {
                                rules.retain(|rule| rule.name() == name_or_category);
                            }
                        }
                    };
                }
            }
        }

        let mut rules = rules.into_iter().collect::<Vec<_>>();
        // for stable diagnostics output ordering
        rules.sort_unstable_by_key(|rule| rule.name());
        rules
    }

    /// # Panics
    ///
    /// * When `mpsc::channel` fails to send.
    #[must_use]
    pub fn run(&self) -> CliRunResult {
        let now = std::time::Instant::now();

        let number_of_files = Arc::new(AtomicUsize::new(0));
        let (tx_error, rx_error) = mpsc::channel::<(PathBuf, Vec<Error>)>();

        self.process_paths(&number_of_files, tx_error);
        let (number_of_warnings, number_of_diagnostics) = self.process_diagnostics(&rx_error);

        CliRunResult::LintResult {
            duration: now.elapsed(),
            number_of_rules: self.linter.number_of_rules(),
            number_of_files: number_of_files.load(Ordering::Relaxed),
            number_of_diagnostics,
            number_of_warnings,
            max_warnings_exceeded: self
                .options
                .max_warnings
                .map_or(false, |max_warnings| number_of_warnings > max_warnings),
        }
    }

    fn process_paths(
        &self,
        number_of_files: &Arc<AtomicUsize>,
        tx_error: mpsc::Sender<(PathBuf, Vec<Error>)>,
    ) {
        let (tx_path, rx_path) = mpsc::channel::<Box<Path>>();

        let walk = Walk::new(&self.options);
        let number_of_files = Arc::clone(number_of_files);
        rayon::spawn(move || {
            let mut count = 0;
            walk.iter().for_each(|path| {
                count += 1;
                tx_path.send(path).unwrap();
            });
            number_of_files.store(count, Ordering::Relaxed);
        });

        let linter = Arc::clone(&self.linter);
        rayon::spawn(move || {
            while let Ok(path) = rx_path.recv() {
                let tx_error = tx_error.clone();
                let linter = Arc::clone(&linter);
                rayon::spawn(move || {
                    if let Some(diagnostics) = Self::lint_path(&linter, &path) {
                        tx_error.send(diagnostics).unwrap();
                    }
                    drop(tx_error);
                });
            }
        });
    }

    fn process_diagnostics(
        &self,
        rx_error: &mpsc::Receiver<(PathBuf, Vec<Error>)>,
    ) -> (usize, usize) {
        let mut number_of_warnings = 0;
        let mut number_of_diagnostics = 0;
        let mut buf_writer = BufWriter::new(std::io::stdout());
        let handler = GraphicalReportHandler::new();

        while let Ok((path, diagnostics)) = rx_error.recv() {
            number_of_diagnostics += diagnostics.len();
            let mut output = String::new();
            for diagnostic in diagnostics {
                if diagnostic.severity() == Some(Severity::Warning) {
                    number_of_warnings += 1;
                    // The --quiet flag follows ESLint's --quiet behavior as documented here: https://eslint.org/docs/latest/use/command-line-interface#--quiet
                    // Note that it does not disable ALL diagnostics, only Warning diagnostics
                    if self.options.quiet {
                        continue;
                    }

                    if let Some(max_warnings) = self.options.max_warnings {
                        if number_of_warnings > max_warnings {
                            continue;
                        }
                    }
                }

                let mut err = String::new();
                handler.render_report(&mut err, diagnostic.as_ref()).unwrap();
                // Skip large output and print only once
                if err.lines().any(|line| line.len() >= 400) {
                    let minified_diagnostic = Error::new(MinifiedFileError(path.clone()));
                    err = format!("{minified_diagnostic:?}");
                    output = err;
                    break;
                }
                output.push_str(&err);
            }
            buf_writer.write_all(output.as_bytes()).unwrap();
        }

        buf_writer.flush().unwrap();
        (number_of_warnings, number_of_diagnostics)
    }

    fn lint_path(linter: &Linter, path: &Path) -> Option<(PathBuf, Vec<Error>)> {
        let source_text = fs::read_to_string(path).unwrap_or_else(|_| panic!("{path:?} not found"));
        let allocator = Allocator::default();
        let source_type =
            SourceType::from_path(path).unwrap_or_else(|_| panic!("incorrect {path:?}"));
        let ret = Parser::new(&allocator, &source_text, source_type).parse();

        if !ret.errors.is_empty() {
            return Some(Self::wrap_diagnostics(path, &source_text, ret.errors));
        };

        let program = allocator.alloc(ret.program);
        let semantic_ret =
            SemanticBuilder::new(&source_text, source_type, &ret.trivias).build(program);

        if !semantic_ret.errors.is_empty() {
            return Some(Self::wrap_diagnostics(path, &source_text, semantic_ret.errors));
        };

        let result = linter.run(&Rc::new(semantic_ret.semantic));

        if result.is_empty() {
            return None;
        }

        if linter.has_fix() {
            let fix_result = Fixer::new(&source_text, result).fix();
            fs::write(path, fix_result.fixed_code.as_bytes()).unwrap();
            let errors = fix_result.messages.into_iter().map(|m| m.error).collect();
            return Some(Self::wrap_diagnostics(path, &source_text, errors));
        }

        let errors = result.into_iter().map(|diagnostic| diagnostic.error).collect();
        Some(Self::wrap_diagnostics(path, &source_text, errors))
    }

    fn wrap_diagnostics(
        path: &Path,
        source_text: &str,
        diagnostics: Vec<Error>,
    ) -> (PathBuf, Vec<Error>) {
        let source = Arc::new(NamedSource::new(path.to_string_lossy(), source_text.to_owned()));
        let diagnostics = diagnostics
            .into_iter()
            .map(|diagnostic| diagnostic.with_source_code(Arc::clone(&source)))
            .collect();
        (path.to_path_buf(), diagnostics)
    }
}
