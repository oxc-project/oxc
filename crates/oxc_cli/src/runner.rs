use std::{
    path::PathBuf,
    process::{ExitCode, Termination},
};

use clap::Command;

pub trait RunnerOptions: for<'a> From<&'a clap::ArgMatches> {
    /// Add CLI arguments to a [`Command`], usually created from a [`Runner`].
    fn build_args(cmd: Command) -> Command;
}

/// A trait for exposing Oxide functionality to the CLI.
pub trait Runner: Send + Sync {
    /// The name of the runner. Used to create a subcommand.
    const NAME: &'static str;
    /// A short description about what the runner does
    const ABOUT: &'static str;
    type Options: RunnerOptions;

    fn command() -> Command {
        let cmd = Command::new(Self::NAME).about(Self::ABOUT);
        Self::Options::build_args(cmd)
    }
    fn new(options: Self::Options) -> Self;

    /// Executes the runner, providing some result to the CLI.
    fn run(&self) -> CliRunResult;
}

#[derive(Debug)]
pub enum CliRunResult {
    None,
    IOError(crate::lint::Error),
    PathNotFound {
        paths: Vec<PathBuf>,
    },
    LintResult {
        duration: std::time::Duration,
        number_of_rules: usize,
        number_of_files: usize,
        number_of_warnings: usize,
        number_of_errors: usize,
        max_warnings_exceeded: bool,
    },
    TypeCheckResult {
        duration: std::time::Duration,
        number_of_diagnostics: usize,
    },
}

impl Termination for CliRunResult {
    fn report(self) -> ExitCode {
        match self {
            Self::None => ExitCode::from(0),
            Self::PathNotFound { paths } => {
                println!("Path {paths:?} does not exist.");
                ExitCode::from(1)
            }
            Self::IOError(e) => {
                println!("IO Error: {e}");
                ExitCode::from(1)
            }
            Self::LintResult {
                duration,
                number_of_rules,
                number_of_files,
                number_of_warnings,
                number_of_errors,
                max_warnings_exceeded,
            } => {
                let ms = duration.as_millis();
                let threads = rayon::current_num_threads();
                let number_of_diagnostics = number_of_warnings + number_of_errors;

                if number_of_diagnostics > 0 {
                    println!();
                }

                println!(
                    "Finished in {ms}ms on {number_of_files} files with {number_of_rules} rules using {threads} threads."
                );

                if max_warnings_exceeded {
                    println!("Exceeded maximum number of warnings. Found {number_of_warnings}.");
                    return ExitCode::from(1);
                }

                if number_of_diagnostics > 0 {
                    let warnings = if number_of_warnings == 1 { "warning" } else { "warnings" };
                    let errors = if number_of_errors == 1 { "error" } else { "errors" };
                    println!(
                        "Found {number_of_warnings} {warnings} and {number_of_errors} {errors}."
                    );
                    return ExitCode::from(1);
                }

                // eslint does not print anything after success, so we do the same.
                // It is also standard to not print anything after success in the *nix world.
                ExitCode::from(0)
            }
            Self::TypeCheckResult { duration, number_of_diagnostics } => {
                let ms = duration.as_millis();
                println!("Finished in {ms}ms.");

                if number_of_diagnostics > 0 {
                    println!("Found {number_of_diagnostics} errors.");
                    return ExitCode::from(1);
                }

                ExitCode::from(0)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use clap::{Arg, ArgAction, ArgMatches, Command};

    use super::*;

    #[derive(Debug)]
    struct TestRunnerOptions {
        foo: bool,
    }

    #[derive(Debug)]
    struct TestRunner {
        options: TestRunnerOptions,
    }

    impl<'a> From<&'a ArgMatches> for TestRunnerOptions {
        fn from(value: &'a ArgMatches) -> Self {
            let foo = value.get_flag("foo");
            Self { foo }
        }
    }

    impl RunnerOptions for TestRunnerOptions {
        fn build_args(cmd: Command) -> Command {
            cmd.arg(Arg::new("foo").short('f').action(ArgAction::SetTrue).required(false))
        }
    }

    impl Runner for TestRunner {
        type Options = TestRunnerOptions;

        const ABOUT: &'static str = "some description";
        const NAME: &'static str = "test";

        fn new(options: Self::Options) -> Self {
            Self { options }
        }

        fn run(&self) -> CliRunResult {
            CliRunResult::None
        }
    }

    #[test]
    fn check_cmd_validity() {
        TestRunner::command().debug_assert();
    }

    #[test]
    fn smoke_test_runner() {
        let cmd = TestRunner::command();
        let matches = cmd.get_matches_from("test -f".split(' '));
        let opts = TestRunnerOptions::from(&matches);
        let runner = TestRunner::new(opts);
        assert!(runner.options.foo);
        let result = runner.run();
        assert!(matches!(result, CliRunResult::None));
    }
}
