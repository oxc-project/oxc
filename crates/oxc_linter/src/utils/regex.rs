use oxc_allocator::Allocator;
use oxc_ast::{AstKind, ast::Argument};
use oxc_regular_expression::{ConstructorParser, Options, ast::Pattern};
use oxc_semantic::IsGlobalReference;
use oxc_span::Span;

use crate::{AstNode, context::LintContext};

pub fn run_on_regex_node<'a, 'b, M>(node: &'a AstNode<'b>, ctx: &'a LintContext<'b>, cb: M)
where
    M: FnOnce(&Pattern<'_>, Span),
{
    match node.kind() {
        AstKind::RegExpLiteral(reg) => {
            if let Some(pat) = &reg.regex.pattern.pattern {
                cb(pat, reg.span);
            }
        }
        AstKind::NewExpression(expr)
            if expr.callee.is_global_reference_name("RegExp", ctx.semantic().scoping()) =>
        {
            // note: improvements required for strings used via identifier references
            // Missing or non-string arguments will be runtime errors, but are not covered by this rule.
            match (&expr.arguments.first(), &expr.arguments.get(1)) {
                (Some(Argument::StringLiteral(pattern)), Some(Argument::StringLiteral(flags))) => {
                    let allocator = Allocator::default();
                    if let Some(pat) = parse_regex(&allocator, pattern.span, Some(flags.span), ctx)
                    {
                        cb(&pat, pattern.span);
                    }
                }
                (Some(Argument::StringLiteral(pattern)), _) => {
                    let allocator = Allocator::default();
                    if let Some(pat) = parse_regex(&allocator, pattern.span, None, ctx) {
                        cb(&pat, pattern.span);
                    }
                }
                _ => {}
            }
        }

        // RegExp()
        AstKind::CallExpression(expr)
            if expr.callee.is_global_reference_name("RegExp", ctx.semantic().scoping()) =>
        {
            // note: improvements required for strings used via identifier references
            // Missing or non-string arguments will be runtime errors, but are not covered by this rule.
            match (&expr.arguments.first(), &expr.arguments.get(1)) {
                (Some(Argument::StringLiteral(pattern)), Some(Argument::StringLiteral(flags))) => {
                    let allocator = Allocator::default();
                    if let Some(pat) = parse_regex(&allocator, pattern.span, Some(flags.span), ctx)
                    {
                        cb(&pat, pattern.span);
                    }
                }
                (Some(Argument::StringLiteral(pattern)), _) => {
                    let allocator = Allocator::default();
                    if let Some(pat) = parse_regex(&allocator, pattern.span, None, ctx) {
                        cb(&pat, pattern.span);
                    }
                }
                _ => {}
            }
        }
        _ => {}
    }
}

fn parse_regex<'a>(
    allocator: &'a Allocator,
    pattern_span: Span,
    flags_span: Option<Span>,
    ctx: &'a LintContext<'_>,
) -> Option<Pattern<'a>> {
    let flags_text = flags_span.map(|span| span.source_text(ctx.source_text()));
    let parser = ConstructorParser::new(
        allocator,
        pattern_span.source_text(ctx.source_text()),
        flags_text,
        Options {
            pattern_span_offset: pattern_span.start,
            flags_span_offset: flags_span.map_or(0, |span| span.start),
        },
    );
    let Ok(pattern) = parser.parse() else { return None };
    Some(pattern)
}
