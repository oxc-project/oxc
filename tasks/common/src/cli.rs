//! Common CLI utilities for task binaries
//!
//! This module provides standardized patterns for CLI argument parsing
//! used across different task binaries in the oxc project.

use pico_args::Arguments;

/// Common CLI options pattern used across multiple tasks
#[derive(Debug, Clone)]
pub struct CommonCliOptions {
    /// Enable debug mode (usually single-threaded execution)
    pub debug: bool,
    /// Filter tests/operations by pattern
    pub filter: Option<String>,
    /// Override existing outputs
    pub r#override: bool,
    /// Show detailed output
    pub detail: bool,
    /// Show diff output
    pub diff: bool,
    /// Execute tests
    pub exec: bool,
}

impl Default for CommonCliOptions {
    fn default() -> Self {
        Self {
            debug: false,
            filter: None,
            r#override: false,
            detail: false,
            diff: false,
            exec: false,
        }
    }
}

impl CommonCliOptions {
    /// Parse common CLI options from command line arguments
    pub fn from_args(args: &mut Arguments) -> Self {
        Self {
            debug: args.contains("--debug"),
            filter: args.opt_value_from_str("--filter").unwrap(),
            r#override: args.contains("--override"),
            detail: args.contains("--detail"),
            diff: args.contains("--diff"),
            exec: args.contains("--exec"),
        }
    }

    /// Create a subset with only debug and filter (most common pattern)
    pub fn debug_filter_only(debug: bool, filter: Option<String>) -> Self {
        Self { debug, filter, ..Default::default() }
    }
}

/// Extract a subcommand from arguments (commonly used pattern)
pub fn get_subcommand(args: &mut Arguments) -> Option<String> {
    args.subcommand().unwrap_or(None)
}

/// Common pattern for handling CLI args in task binaries
pub fn parse_common_args() -> (Option<String>, CommonCliOptions) {
    let mut args = Arguments::from_env();
    let subcommand = get_subcommand(&mut args);
    let options = CommonCliOptions::from_args(&mut args);
    (subcommand, options)
}
