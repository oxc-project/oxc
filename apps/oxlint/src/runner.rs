use crate::cli::CliRunResult;

/// A trait for exposing functionality to the CLI.
pub trait Runner {
    type Options;

    fn new(matches: Self::Options) -> Self;

    /// Executes the runner, providing some result to the CLI.
    fn run(self) -> CliRunResult;
}
