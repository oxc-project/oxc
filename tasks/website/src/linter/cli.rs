use oxlint::cli::lint_command;

#[test]
fn test_cli() {
    let snapshot = generate_cli();
    insta::with_settings!({ prepend_module_to_snapshot => false }, {
        insta::assert_snapshot!(snapshot);
    });
}

#[test]
fn test_cli_terminal() {
    let snapshot = oxlint::cli::lint_command().run_inner(&["--help"]).unwrap_err().unwrap_stdout();
    insta::with_settings!({ prepend_module_to_snapshot => false }, {
        insta::assert_snapshot!(snapshot);
    });
}

// <https://oxc.rs/docs/guide/usage/linter/cli.html>
pub fn print_cli() {
    println!("{}", generate_cli());
}

fn generate_cli() -> String {
    let markdown = lint_command().render_markdown("oxlint");
    // Remove the extra header
    let markdown = markdown.trim_start_matches("# oxlint\n");

    // Hack usage line
    let markdown = markdown.replacen("**Usage**:", "## Usage\n", 1);

    let markdown = markdown
        .split('\n')
        .flat_map(|line| {
            // Hack the bug on the line containing `###`
            if line.contains("###") {
                line.split("###").map(str::trim).chain(["\n"]).collect::<Vec<_>>()
            } else {
                vec![line]
            }
        })
        .map(|line| {
            // Make `** title **` to `## title`
            if let Some(line) = line.strip_prefix("**") {
                if let Some(line) = line.strip_suffix("**") {
                    return format!("## {line}");
                }
            }

            line.to_string()
        })
        .collect::<Vec<_>>()
        .join("\n");

    markdown
}
