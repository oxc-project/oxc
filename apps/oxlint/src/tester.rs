#[cfg(test)]
use crate::cli::{lint_command, CliRunResult, LintResult, LintRunner};
#[cfg(test)]
use crate::runner::Runner;
#[cfg(test)]
use regex::Regex;
#[cfg(test)]
use std::{env, path::PathBuf};
#[cfg(test)]
pub struct Tester {
    cwd: PathBuf,
}

#[cfg(test)]
impl Tester {
    pub fn new() -> Self {
        let cwd = env::current_dir().unwrap();

        Self { cwd }
    }

    pub fn with_cwd(mut self, cwd: PathBuf) -> Self {
        self.cwd.push(cwd);
        self
    }

    pub fn get_lint_result(&self, args: &[&str]) -> LintResult {
        let mut new_args = vec!["--silent"];
        new_args.extend(args);

        let options = lint_command().run_inner(new_args.as_slice()).unwrap();
        let mut output = Vec::new();
        match LintRunner::new(options).with_cwd(self.cwd.clone()).run(&mut output) {
            CliRunResult::LintResult(lint_result) => lint_result,
            other => panic!("{other:?}"),
        }
    }

    pub fn get_invalid_option_result(&self, args: &[&str]) -> String {
        let mut new_args = vec!["--silent"];
        new_args.extend(args);

        let options = lint_command().run_inner(new_args.as_slice()).unwrap();
        let mut output = Vec::new();
        match LintRunner::new(options).with_cwd(self.cwd.clone()).run(&mut output) {
            CliRunResult::InvalidOptions { message } => message,
            other => {
                panic!("Expected InvalidOptions, got {other:?}");
            }
        }
    }

    pub fn test_and_snapshot(&self, args: &[&str]) {
        let mut settings = insta::Settings::clone_current();

        let options = lint_command().run_inner(args).unwrap();
        let mut output: Vec<u8> = Vec::new();
        let args_string = args.join(" ");

        output.extend_from_slice(format!("########## \n{args_string}\n----------\n").as_bytes());
        let _ = LintRunner::new(options).with_cwd(self.cwd.clone()).run(&mut output);
        output.push(b'\n');

        settings.set_prepend_module_to_snapshot(false);
        settings.set_omit_expression(true);
        settings.set_snapshot_suffix("oxlint");

        let regex = Regex::new(r"\d+ms|\d+ threads?").unwrap();

        let output_string = &String::from_utf8(output).unwrap();
        let output_string = regex.replace_all(output_string, "<variable>");

        settings.bind(|| {
            insta::assert_snapshot!(format!("{}", args_string), output_string);
        });
    }
}
