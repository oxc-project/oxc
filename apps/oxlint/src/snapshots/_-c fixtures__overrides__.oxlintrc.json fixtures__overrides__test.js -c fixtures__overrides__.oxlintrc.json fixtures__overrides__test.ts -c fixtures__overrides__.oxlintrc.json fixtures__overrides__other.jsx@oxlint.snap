---
source: apps/oxlint/src/tester.rs
---
########## 
arguments: -c fixtures/overrides/.oxlintrc.json fixtures/overrides/test.js
working directory: 
----------

  x ]8;;https://oxc.rs/docs/guide/usage/linter/rules/eslint/no-var.html\eslint(no-var)]8;;\: Unexpected var, use let or const instead.
   ,-[fixtures/overrides/test.js:1:1]
 1 | var msg = "hello";
   : ^^^
 2 | console.log(msg);
   `----
  help: Replace var with let or const

Found 0 warnings and 1 error.
Finished in <variable>ms on 1 file with 99 rules using 1 threads.
----------
CLI result: LintFoundErrors
----------

########## 
arguments: -c fixtures/overrides/.oxlintrc.json fixtures/overrides/test.ts
working directory: 
----------

  x ]8;;https://oxc.rs/docs/guide/usage/linter/rules/eslint/no-var.html\eslint(no-var)]8;;\: Unexpected var, use let or const instead.
   ,-[fixtures/overrides/test.ts:1:1]
 1 | var msg = "hello";
   : ^^^
 2 | console.log(msg);
   `----
  help: Replace var with let or const

  ! ]8;;https://oxc.rs/docs/guide/usage/linter/rules/eslint/no-console.html\eslint(no-console)]8;;\: eslint(no-console): Unexpected console statement.
   ,-[fixtures/overrides/test.ts:2:1]
 1 | var msg = "hello";
 2 | console.log(msg);
   : ^^^^^^^^^^^
   `----
  help: Delete this console statement.

Found 1 warning and 1 error.
Finished in <variable>ms on 1 file with 99 rules using 1 threads.
----------
CLI result: LintFoundErrors
----------

########## 
arguments: -c fixtures/overrides/.oxlintrc.json fixtures/overrides/other.jsx
working directory: 
----------

  x ]8;;https://oxc.rs/docs/guide/usage/linter/rules/eslint/no-var.html\eslint(no-var)]8;;\: Unexpected var, use let or const instead.
   ,-[fixtures/overrides/other.jsx:1:1]
 1 | var msg = "hello";
   : ^^^
 2 | console.log(msg);
   `----
  help: Replace var with let or const

Found 0 warnings and 1 error.
Finished in <variable>ms on 1 file with 99 rules using 1 threads.
----------
CLI result: LintFoundErrors
----------
